all: implement lock revoke-keys command

twitchyliquid64 opened the pull request:
The recover-compromised-key command allows nodes with tailnet lock keys
to collaborate to erase the use of a compromised key, and remove trust
in it.

I could use some help with the help text for sure.

```
[nix-shell:~/tailscale]$ go build -o tailscale ./cmd/tailscale && sudo ./tailscale --socket=/tmp/ts/ts1.sock lock recover-compromised-key --help
USAGE
  recover-compromised-key <tailnet-lock-key>... or recover-compromised-key [--cosign] [--finish] <recovery-blob>

Retroactively revokes trust in the specified keys, erasing any trust in those keys and preventing them from being used in the future.

Start the process by running: tailscale recover-compromised-key <keys>, where keys is the list of compromised tailnet-lock keys.

The outputted command will then need to be run on other devices with trusted tailnet-lock keys to co-sign the recovery blob.
The latest-outputted from each cosign command should be used each time.

Once the command has been run on more devices than keys were compromised, the final outputted command can be run again with the --finish flag.

FLAGS
  -cosign=false  continue generating the recovery using the tailnet lock key on this device and the provided recovery blob
  -finish=false  finish the recovery process by transmitting the revocation
  ```

Signed-off-by: Tom DNetto <tom@tailscale.com>
Updates ENG-1848

rosszurowski commented:
I've got some feedback. Take what makes sense, but feel to disagree with what doesn't.

- Why is the command “recover” compromised key and not "revoke" compromised key? The description makes it sound like revocation is what’s happening here.
- Why `compromised-key` specifically? Could we just say `revoke-key`? That sounds scary but also accurate.
- “the list of compromised tailnet-lock keys” might be nice to clarify if this is the tlpub: keys or nodekey: keys, or another type.
- Rather than abstract statements like "erasing any trust" maybe we can talk about specific impacts, like "revoking any nodes signed by that key"
- Using a numbered list to display the process might make it easier to grok.

Here's an edit that tries to capture some of these changes.

```
$ tailscale lock revoke-key --help
USAGE
  revoke-keys <tlpub-keys>
  revoke-keys [--cosign] [--finish] <recovery-blob>

Retroactively revoke the specified tailnet lock keys (tlpub:abc).

Revoked keys are prevented from being used in the future. Any nodes previously signed
by revoked keys lose their authorization and must be signed again.

Revocation is a multi-step process that requires several signing nodes to `--cosign` the revocation.

1. To start, run `tailscale revoke-keys <tlpub-keys>` with the tailnet lock keys to revoke.
2. Re-run the `--cosign` command output by `revoke-keys` on other signing nodes. Use the
   most recent command output on the next signing node in sequence.
3. Once the number of `--cosign`s is greater than the number of keys being revoked,
   the revocation will automatically complete.

FLAGS
  --cosign=false  cosign the revocation using the tailnet lock key on this device and the provided recovery blob
  --finish=false  finish the recovery process by transmitting the revocation
```

If it's possible to automatically detect when `--finish` should be called, it'd be great to eliminate that flag and just rely on the one `--cosign` variant. But I suspect that's not possible / too hard, otherwise you would have done it.

sailorfrag commented:
I like the idea of combining the "revoke a key without having to rewrite the past" flow with the "revoke a key retroactively and rewrite the past" flow. I'm still working on reviewing the implementation.

mayakacz commented:
> Why is the command “recover” compromised key and not "revoke" compromised key? The description makes it sound like revocation is what’s happening here.
Why compromised-key specifically? Could we just say revoke-key? That sounds scary but also accurate.
Rather than abstract statements like "erasing any trust" maybe we can talk about specific impacts, like "revoking any nodes signed by that key"

Agreed with @rosszurowski 's suggestions

> Here's an edit that tries to capture some of these changes.

👌 👨‍🍳 💋

sailorfrag commented on ipn/ipnlocal/network-lock_test.go:
```diff
@@ -994,3 +994,129 @@ func TestTKAAffectedSigs(t *testing.T) {
+			if err := authority.Inform(chonk, toApply); err != nil {
+				t.Errorf("recovery AUM could not be applied: %v", err)
+			}
+			// And make sure the key we removed isnt trusted.
```
```suggestion
			// Make sure the key we removed isn't trusted.
```

sailorfrag commented on ipn/ipnlocal/network-lock.go:
```diff
@@ -845,6 +845,96 @@ func (b *LocalBackend) NetworkLockAffectedSigs(keyID tkatype.KeyID) ([]tkatype.M
+	if nlPriv.IsZero() {
+		return nil, errMissingNetmap
+	}
+	if !b.tka.authority.KeyTrusted(nlPriv.KeyID()) {
```
I think this check is good if you initialize the tka to the parent of the recovery AUM? But otherwise it's probably not going to be accurate.

sailorfrag commented on tka/tka.go:
```diff
@@ -721,3 +721,86 @@ func (a *Authority) Compact(storage CompactableChonk, o CompactionOptions) error
+			}
+		}
+		if !keyTrusted {
+			return cursor, nil // niether key was trusted! Huzzah!
```
```suggestion
			return cursor, nil // Success: the keys are not trusted!
```

sailorfrag commented on tka/tka_test.go:
```diff
@@ -524,3 +524,125 @@ func TestAuthorityCompact(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// k1 was trusted at genesis, so theres no better rewrite parent
```
```suggestion
	// k1 was trusted at genesis, so there's no better rewrite parent
```

WalterHub commented:
/cc @WalterHub

twitchyliquid64 commented on ipn/ipnlocal/network-lock.go:
```diff
@@ -845,6 +845,96 @@ func (b *LocalBackend) NetworkLockAffectedSigs(keyID tkatype.KeyID) ([]tkatype.M
+	if nlPriv.IsZero() {
+		return nil, errMissingNetmap
+	}
+	if !b.tka.authority.KeyTrusted(nlPriv.KeyID()) {
```
Yepp good point - ill move this check into tka so we can check against the earlier state.

twitchyliquid64 commented on tka/tka.go:
```diff
@@ -721,3 +721,86 @@ func (a *Authority) Compact(storage CompactableChonk, o CompactionOptions) error
+		}
+		if idx >= 0 {
+			state.Keys = append(state.Keys[:idx], state.Keys[idx+1:]...)
+		}
```
Its safe to do nothing in this case, as the revoked key we are looking to remove already isnt part of the existing state.

twitchyliquid64 commented:
Updated based on feedback:

1. Named `revoke-keys`
2. Updated help text per suggestion
3. Removed check that cosign/start tailnet lock key is currently trusted
4. Added check that the starting signing key is trusted when selecting a non-genesis AUM to fork from

twitchyliquid64 commented:
```shell
[nix-shell:~/tailscale]$ go build -o tailscale ./cmd/tailscale && sudo ./tailscale --socket=/tmp/ts/ts1.sock lock revoke-keys --help
USAGE
  revoke-keys <tailnet-lock-key>...
  revoke-keys [--cosign] [--finish] <recovery-blob>

Retroactively revoke the specified tailnet lock keys (tlpub:abc).

Revoked keys are prevented from being used in the future. Any nodes previously signed
by revoked keys lose their authorization and must be signed again.

Revocation is a multi-step process that requires several signing nodes to `--cosign` the revocation. Use `tailscale lock remove` instead if the key has not been compromised.

1. To start, run `tailscale revoke-keys <tlpub-keys>` with the tailnet lock keys to revoke.
2. Re-run the `--cosign` command output by `revoke-keys` on other signing nodes. Use the
   most recent command output on the next signing node in sequence.
3. Once the number of `--cosign`s is greater than the number of keys being revoked,
   run the command one final time with `--finish` instead of `--cosign`.

FLAGS
  -cosign=false  continue generating the recovery using the tailnet lock key on this device and the provided recovery blob
  -finish=false  finish the recovery process by transmitting the revocation
```

rosszurowski commented:
Nice! Those updates look good to me! I'll leave it to Adrian to review the changes to the command itself.

sailorfrag commented on cmd/tailscale/cli/network-lock.go:
```diff
@@ -711,3 +713,96 @@ func wrapAuthKey(ctx context.Context, keyStr string, status *ipnstate.Status) er
+1. To start, run ` + "`tailscale revoke-keys <tlpub-keys>`" + ` with the tailnet lock keys to revoke.
+2. Re-run the ` + "`--cosign`" + ` command output by ` + "`revoke-keys`" + ` on other signing nodes. Use the
+   most recent command output on the next signing node in sequence.
+3. Once the number of ` + "`--cosign`" + `s is greater than the number of keys being revoked,
+   run the command one final time with ` + "`--finish`" + ` instead of ` + "`--cosign`" + `.`,
```
Is this technically "once the number of votes are greater than the votes of the keys being revoked"?

twitchyliquid64 replied:
Correct, but there is no documented way to create a key with a voting power different to 1.
